st2actions/st2actions/scheduler/handler.py

```diff
@@ -0,0 +1,164 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+import eventlet
+
+from st2common import log as logging
+from st2common.util import date
+from st2common.constants import action as action_constants
+from st2common.exceptions.db import StackStormDBObjectNotFoundError
+from st2common.models.db.liveaction import LiveActionDB
+from st2common.services import action as action_service
+from st2common.services import policies as policy_service
+from st2common.persistence.liveaction import LiveAction
+from st2common.persistence.execution_queue import ExecutionQueue
+from st2common.util import action_db as action_utils
+from st2common.services.coordination import LockAcquireError
+from st2common.services import coordination
+
+__all__ = [
+    'ExecutionQueueHandler',
+    'get_handler'
+]
+
+
+LOG = logging.getLogger(__name__)
+
+
+def _next_executions():
+    """
+        Sort executions by fifo and priority and get the latest, highest priority
+        item from the queue and pop it off.
+    """
+    query = {
+        "scheduled_start_timestamp__lte": date.get_datetime_utc_now(),
+        "order_by": [
+            "scheduled_start_timestamp",
+        ]
+    }
+
+    queued_executions = ExecutionQueue.query(**query)
+    if queued_executions:
+        return queued_executions
+
+    return []
+
+
+class ExecutionQueueHandler(object):
+    def __init__(self):
+        self.message_type = LiveActionDB
+        self._shutdown = False
+
+    def loop(self):
+        LOG.debug('Entering scheduler loop')
+        while self._shutdown is not True:
+            eventlet.greenthread.sleep(0.25)
+            try:
+                with coordination.lock('st2shcedulerqueue'):
```

Comment:
This is not the most ideal lock since it's quite high level and limits concurrent processing of executions.

For now though, it should probably work, but eventually we should figure out if we can make it more fine granular so we have multiple scheduler services processing set of executions which are pending to be processed.